google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java

```diff
@@ -564,6 +579,89 @@ public void keepAlive() throws Exception {
   }
 
+  @Test
+  public void blockAndTimeoutOnPoolExhaustion() throws InterruptedException, ExecutionException {
+    if (minSessions != 0) {
+      // Only execute for minSessions == 0 as we need to setup and shutdown a mock sampler during
+      // the test case. A second run would not work as there is no way to restart the sampler.
+      return;
+    }
+    // Setup a dummy trace handler.
+    final AtomicInteger deadlineExceededCount = new AtomicInteger();
+    Handler handler =
+        new Handler() {
+          @Override
+          public void export(Collection<SpanData> spanDataList) {
+            for (SpanData sd : spanDataList) {
+              if (sd.getStatus() == Status.DEADLINE_EXCEEDED) {
+                deadlineExceededCount.incrementAndGet();
+              }
+            }
+          }
+        };
+    Tracing.getExportComponent()
+        .getSpanExporter()
+        .registerHandler(SessionPoolTest.class.getName(), handler);
+    TraceConfig traceConfig = Tracing.getTraceConfig();
+    TraceParams activeTraceParams = traceConfig.getActiveTraceParams();
+    traceConfig.updateActiveTraceParams(
+        activeTraceParams.toBuilder().setSampler(Samplers.alwaysSample()).build());
+    // Create a session pool with max 1 session and a low timeout for waiting for a session.
+    options =
+        SessionPoolOptions.newBuilder()
+            .setMinSessions(minSessions)
+            .setMaxSessions(1)
+            .setInitialWaitForSessionTimeoutMillis(20L)
+            .build();
+    SessionImpl mockSession = mockSession();
+    when(client.createSession(db)).thenReturn(mockSession);
+    pool = createPool();
+
+    Tracer tracer = Tracing.getTracer();
+    Span span = tracer.spanBuilder("RunTest").startSpan();
+    // Try to take a read or a read/write session. These requests should block.
+    for (Boolean write : new Boolean[] {true, false}) {
+      // Take the only session that can be in the pool.
+      Session checkedOutSession = pool.getReadSession();
+      final Boolean finWrite = write;
+      ExecutorService executor = Executors.newFixedThreadPool(1);
+      Future<Void> fut =
+          executor.submit(
+              new Callable<Void>() {
+                @Override
+                public Void call() throws Exception {
+                  Session session;
+                  if (finWrite) {
+                    session = pool.getReadWriteSession();
+                  } else {
+                    session = pool.getReadSession();
+                  }
+                  session.close();
+                  return null;
+                }
+              });
+      try {
+        fut.get(40L, TimeUnit.MILLISECONDS);
+        fail("missing expected timeout exception");
+      } catch (TimeoutException e) {
+        // Return the checked out session to the pool so the async request will get a session and
+        // finish.
+        checkedOutSession.close();
+        fut.get();
+        executor.shutdown();
```

Comment:
close() and shutdown() should be done in a finally as we should be cleaning up regardless of whether the timeout happened or not.

Reply:
Changed.